### PR TITLE
bpo-28837: Fix lib2to3 handling of map/zip/filter

### PR DESCRIPTION
**Pull request by stuarteberg**
The `2to3` tool will automatically wrap calls to `map`, `zip`, and `filter` with `list()`:

```python
# Original:
zip('abc', '123')

# After 2to3:
list(zip('abc', '123'))
```

...but the tool misses some cases, if the call is followed by a "trailer" such as `[0]`:

```python
# Skipped by 2to3:
zip('abc', '123')[0]
```

This PR augments the `lib2to3` "fixers" for `map`, `zip`, and `filter` to correctly handle such cases.

http://bugs.python.org/issue28837

### CODE REVIEW

**Comment by stuarteberg** [comment]
@benjaminp: Are you the right person to ping about `lib2to3` PRs?

File: Lib/lib2to3/fixes/fix_map.py [hunk 1]
```diff
@@ -47,10 +50,12 @@ class FixMap(fixer_base.ConditionalFix):
             >
             ')'
         >
+        [extra_trailers=trailer*]
     >
     |
     power<
-        'map' trailer< '(' [arglist=any] ')' >
+        'map' args=trailer< '(' [any] ')' >
```
Comment (benjaminp) [inline Lib/lib2to3/fixes/fix_map.py:57]: What's the reason for this change?

Comment (stuarteberg) [reply Lib/lib2to3/fixes/fix_map.py:57]: >What's the reason for this change?

On line 100 [(below)](https://github.com/python/cpython/pull/24/files#diff-cea1b4282c04d759aea2a060e04611b5R100), I need to clone the argument list:

```python
new = Node(syms.power, [Name("map"), args.clone()])
```

If I leave the above line unchanged, then I could have written this, instead:

```python
new = Node(syms.power, [Name("map"), ArgList([args.clone()])])
```

...but there's a problem with that:  It doesn't clone the `(` and `)` tokens -- it creates them from scratch.  As a result, one of the tests fails, due to an unpreserved `prefix` before the `)` token:

```
======================================================================
FAIL: test_prefix_preservation (Lib.lib2to3.tests.test_fixers.Test_map)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bergs/Documents/workspace/cpython-git/Lib/lib2to3/tests/test_fixers.py", line 3036, in test_prefix_preservation
    self.check(b, a)
  File "/Users/bergs/Documents/workspace/cpython-git/Lib/lib2to3/tests/test_fixers.py", line 3031, in check
    super(Test_map, self).check(b, a)
  File "/Users/bergs/Documents/workspace/cpython-git/Lib/lib2to3/tests/test_fixers.py", line 36, in check
    tree = self._check(before, after)
  File "/Users/bergs/Documents/workspace/cpython-git/Lib/lib2to3/tests/test_fixers.py", line 32, in _check
    self.assertEqual(after, str(tree))
AssertionError: "x =    list(map(   f,    'abc'   ))\n\n" != "x =    list(map(   f,    'abc'))\n\n"
- x =    list(map(   f,    'abc'   ))
?                               ---
+ x =    list(map(   f,    'abc'))
```

Hence, I think it's best to capture and clone the whole `trailer`.  But if I'm missing something, please let me know.

Comment (benjaminp) [reply Lib/lib2to3/fixes/fix_map.py:57]: Okay, thanks for the explanation.
